extensions/lifecycle/tasks/LifecycleColdStatusArchiveTask.js

```diff
@@ -0,0 +1,75 @@
+const async = require('async');
+
+const ObjectMDArchive = require('arsenal').models.ObjectMDArchive;
+const LifecycleUpdateTransitionTask = require('./LifecycleUpdateTransitionTask');
+
+class LifecycleColdStatusArchiveTask extends LifecycleUpdateTransitionTask {
+    /**
+     * Process a lifecycle object entry
+     *
+     * @constructor
+     * @param {LifecycleObjectProcessor} proc - object processor instance
+     */
+    constructor(proc) {
+        const procState = proc.getStateVars();
+        super();
+        Object.assign(this, procState);
+    }
+
+    getTargetAttribute(entry) {
+        const {
+            bucketName,
+            objectKey,
+            objectVersion,
+            accountId
+        } = entry.getTarget();
+        return {
+            bucket: bucketName,
+            key: objectKey,
+            version: objectVersion,
+            accountId
+        };
+    }
+
+    getSourceAttribute(entry) {
+        const { bucketName, objectKey, objectVersion } = entry.getTarget();
+        return {
+            bucket: bucketName,
+            objectKey,
+            version: objectVersion,
+            storageClass: entry.get('sourceStorageClass'),
+            lastModified: entry.get('sourceLastModified'),
+        };
+    }
+
+    _updateMdWithArchiveInfo(entry, objMD) {
+        const archiveInfo = entry.getArchiveInfo();
+        objMD.setArchive(new ObjectMDArchive(archiveInfo));
+        // set location to empty
+        objMD.setLocation();
+    }
+
+    processEntry(entry, done) {
+        const log = this.logger.newRequestLogger();
+        let locationToGC;
+        return async.waterfall([
+            next => this._getMetadata(entry, log, next),
+            (objMD, next) => {
+                locationToGC = objMD.getLocation();
+                entry.set('sourceStorageClass', objMD.getDataStoreName());
+                entry.set('sourceLastModified', objMD.getLastModified());
+                this._updateMdWithArchiveInfo(entry, objMD);
+                return this._putMetadata(entry, objMD, log, next);
+            },
+            next => {
```

Comment:
This sequence can create orphans if `_garbageCollectLocation` fails (retry won't find any location to gc), and `_garbageCollectLocation` seems to ignore errors.

Reply:
update:
* add delete method without relying on existing GC processor
* split the update md into 2 parts; archive update and location removal update
* new sequence:
** get objectMD
** update objectMD archive info and write to db
** perform location delete
*** on success, unset objectMD location and write to db
*** on error, return error and retry